src/metabase/util/schema.clj

```diff
@@ -399,6 +609,39 @@
                            s/Keyword                              s/Any}
     (deferred-tru "parameter must be a map with :id and :type keys")))
 
+(defschema ParameterSourceMalli
+  [:multi {:dispatch :values_source_type}
+   ["card"        [:map
+                   [:values_source_type :string]
+                   [:values_source_config
+                    [:map {:closed true}
+                     [:card_id {:optional true} IntGreaterThanZeroMalli]
+                     [:value_field {:optional true} FieldMalli]
+                     [:label_field {:optional true} FieldMalli]]]]]
+   ["static-list" [:map
+                   [:values_source_type :string]
+                   [:values_source_config
+                    [:map {:closed true}
+                     [:values {:optional true} [:* :any]]]]]]])
+
+(defschema ParameterMalli
+  "Schema for a valid Parameter.
+  We're not using [metabase.mbql.schema/Parameter] here because this Parameter is meant to be used for
+  Parameters we store on dashboard/card, and it has some difference with Parameter in MBQL."
+  ;; TODO we could use :multi to dispatch values_source_type to the correct values_source_config
+  (with-error-message-malli
+    [:map
+     [:id NonBlankStringMalli]
+     [:type keyword-or-non-blank-str-malli]
+     ;; TODO how to merge this with ParameterSourceMalli above?
```

Comment:
best to merge these with each of the :multi's children's maps

Reply:
ha. Thanks for letting me know.